File: app/web/lib/book-info.ts
import type { BookInfoItem } from '@/lib/types'

type TFn = (key: string, vars?: Record<string, string | number>) => string

// 书籍「更多信息」的预置类型（与服务端 app.bookInfoTypes 一致）：图标、取值类型（决定校验与展示方式）。
export const BOOK_INFO_TYPES: { type: string; icon: string; kind: 'url' | 'email' | 'text' }[] = [
  { type: 'github', icon: 'fa-brands fa-github', kind: 'url' },
  { type: 'gitlab', icon: 'fa-brands fa-gitlab', kind: 'url' },
  { type: 'gitee', icon: 'fa-solid fa-code-branch', kind: 'url' },
  { type: 'website', icon: 'fa-solid fa-globe', kind: 'url' },
  { type: 'source', icon: 'fa-solid fa-file-lines', kind: 'url' },
  { type: 'docs', icon: 'fa-solid fa-book', kind: 'url' },
  { type: 'demo', icon: 'fa-solid fa-display', kind: 'url' },
  { type: 'email', icon: 'fa-solid fa-envelope', kind: 'email' },
  { type: 'author', icon: 'fa-solid fa-user-pen', kind: 'text' },
  { type: 'license', icon: 'fa-solid fa-scale-balanced', kind: 'text' },
  { type: 'version', icon: 'fa-solid fa-tag', kind: 'text' },
  { type: 'isbn', icon: 'fa-solid fa-barcode', kind: 'text' },
  { type: 'custom', icon: 'fa-solid fa-circle-info', kind: 'text' },
]

export function bookInfoType(type: string) {
  return BOOK_INFO_TYPES.find((d) => d.type === type) || BOOK_INFO_TYPES[BOOK_INFO_TYPES.length - 1]
}

const isHTTP = (v: string) => /^https?:\/\/\S+$/i.test(v)

// bookInfoLabel 显示名：自定义名称优先，否则按类型取默认名（i18n：bookInfo.type.<type>）。
export function bookInfoLabel(t: TFn, item: BookInfoItem): string {
  return item.label?.trim() || t(`bookInfo.type.${bookInfoType(item.type).type}`)
}

// bookInfoHref 可点击的链接：链接类型/自定义项中的 http(s) 值 → 原链接；邮箱 → mailto；其余不可点击。
export function bookInfoHref(item: BookInfoItem): string | undefined {
  const kind = bookInfoType(item.type).kind
  if (kind === 'email') return `mailto:${item.value}`
  if ((kind === 'url' || item.type === 'custom') && isHTTP(item.value)) return item.value
  return undefined
}

// bookInfoDisplay 链接类展示时去掉协议与末尾斜杠（完整地址在 Tooltip 中）。
export function bookInfoDisplay(item: BookInfoItem): string {
  return bookInfoHref(item) && isHTTP(item.value) ? item.value.replace(/^https?:\/\//i, '').replace(/\/$/, '') : item.value
}

File: server/internal/app/book_info.go
package app

import (
	"fmt"
	"net/mail"
	"net/url"
	"strings"
	"unicode/utf8"

	"knowforge/server/internal/models"
)

// 书籍「更多信息」：作者为书籍添加附加属性（GitHub 仓库、原始文档地址、许可证等），在书籍详情页展示。
// 预置类型决定前端图标、默认名称与取值校验；custom 为自定义项（需填写名称，值为链接时展示为链接）。

const (
	maxBookInfoItems = 20
	maxBookInfoLabel = 40
	maxBookInfoValue = 500
)

// 取值类型：url 需为 http(s) 链接；email 需为邮箱；text 任意文本。
var bookInfoTypes = map[string]string{
	"github": "url", "gitlab": "url", "gitee": "url", "website": "url", "source": "url", "docs": "url", "demo": "url",
	"email": "email", "license": "text", "author": "text", "version": "text", "isbn": "text", "custom": "text",
}

func isHTTPURL(v string) bool {
	u, err := url.Parse(v)
	return err == nil && (u.Scheme == "http" || u.Scheme == "https") && u.Host != ""
}

// normalizeBookInfo 校验并规范化附加属性（去空白、丢弃空值项）；不合法时返回可展示的错误。
func normalizeBookInfo(items []models.BookInfoItem) (models.BookInfo, error) {
	out := models.BookInfo{}
	for i, it := range items {
		it.Type = strings.TrimSpace(it.Type)
		it.Label = strings.TrimSpace(it.Label)
		it.Value = strings.TrimSpace(it.Value)
		if it.Value == "" {
			continue
		}
		kind, known := bookInfoTypes[it.Type]
		if !known {
			return nil, fmt.Errorf("第 %d 项的类型无效", i+1)
		}
		if it.Type == "custom" && it.Label == "" {
			return nil, fmt.Errorf("第 %d 项为自定义属性，请填写名称", i+1)
		}
		if utf8.RuneCountInString(it.Label) > maxBookInfoLabel {
			return nil, fmt.Errorf("第 %d 项的名称不能超过 %d 个字符", i+1, maxBookInfoLabel)
		}
		if utf8.RuneCountInString(it.Value) > maxBookInfoValue {
			return nil, fmt.Errorf("第 %d 项的内容不能超过 %d 个字符", i+1, maxBookInfoValue)
		}
		switch kind {
		case "url":
			if !isHTTPURL(it.Value) {
				return nil, fmt.Errorf("第 %d 项需为 http(s) 链接", i+1)
			}
		case "email":
			if addr, err := mail.ParseAddress(it.Value); err != nil || addr.Address != it.Value {
				return nil, fmt.Errorf("第 %d 项需为有效的邮箱地址", i+1)
			}
		}
		out = append(out, it)
	}
	if len(out) > maxBookInfoItems {
		return nil, fmt.Errorf("更多信息最多 %d 项", maxBookInfoItems)
	}
	return out, nil
}
